### PR TITLE
Add a HighFive reactor for slash commands and in-chat messages

### PR DESCRIPTION
**Pull request by wolfsage**

### CODE REVIEW

File: lib/Synergy/Reactor/HighFive.pm [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    }
+  );
+
+  return $res;
+};
+
+sub listener_specs {
+  return {
+    name      => 'highfive',
+    method    => 'high_five',
```
Comment (mmcclimon) [inline lib/Synergy/Reactor/HighFive.pm]: `highfive` vs `high_five` seems silly and arbitrary, though I totally understand why you did it.

File: lib/Synergy/Reactor/HighFive.pm [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  };
+}
+
+sub start ($self) {}
+
+sub high_five ($self, $event) {
+  $event->mark_handled;
+
+  unless ($event->from_user) {
+    $event->reply("Sorry, I don't know who you are.");
```
Comment (mmcclimon) [inline lib/Synergy/Reactor/HighFive.pm]: Needs a `return`.

File: lib/Synergy/Reactor/HighFive.pm [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  return $res;
+};
+
+sub listener_specs {
+  return {
+    name      => 'highfive',
+    method    => 'high_five',
+    predicate => sub ($, $e) {
+      $e->was_targeted &&
+      $e->text =~ /^highfive.*$/i;
```
Comment (mmcclimon) [inline lib/Synergy/Reactor/HighFive.pm]: (nit) This `.*` is unnecessary

File: lib/Synergy/Reactor/HighFive.pm [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  my $success = $arg{success} || sub {};
+  my $failure = $arg{failure} || sub {};
+
+  # "highfive to rjbs for eating pie" or "highfive for @rjbs for baking pie"
+  $text =~ s/\Ahighfive\s*(to|for)\s+//i;
+
+  my ($target, $reason) = split /\s+/, $text, 2;
+
+  my $target_user = $self->hub->user_directory->user_named($target);
+  unless ($target_user) {
```
Comment (mmcclimon) [inline lib/Synergy/Reactor/HighFive.pm]: This is totally fine, but I think current highfive behavior allows you to send a highfive to anything, and new behavior won't.

That is, you can say `/highfive @phl for being great` or `/highfive tests for existing`

File: lib/Synergy/Reactor/HighFive.pm [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+sub do_high_five ($self, %arg) {
+  my $from = $arg{from};
+  my $text = $arg{text};
+  my $chan = $arg{chan} || '';
+  my $success = $arg{success} || sub {};
+  my $failure = $arg{failure} || sub {};
+
+  # "highfive to rjbs for eating pie" or "highfive for @rjbs for baking pie"
```
Comment (mmcclimon) [inline lib/Synergy/Reactor/HighFive.pm]: I feel like `@synergy highfive rjbs: eating pie` should work, without the "to" or "for"

File: lib/Synergy/Reactor/HighFive.pm (hunk 1 above)
Comment (wolfsage) [reply lib/Synergy/Reactor/HighFive.pm]: I meant to standardize. Can probably just do "highfive" and be done with it

File: lib/Synergy/Reactor/HighFive.pm (hunk 2 above)
Comment (wolfsage) [reply lib/Synergy/Reactor/HighFive.pm]: Right right thanks

File: lib/Synergy/Reactor/HighFive.pm (hunk 3 above)
Comment (wolfsage) [reply lib/Synergy/Reactor/HighFive.pm]: Oops

File: lib/Synergy/Reactor/HighFive.pm (hunk 4 above)
Comment (wolfsage) [reply lib/Synergy/Reactor/HighFive.pm]: Oh good call

File: lib/Synergy/Reactor/HighFive.pm (hunk 5 above)
Comment (wolfsage) [reply lib/Synergy/Reactor/HighFive.pm]: It does. This regex is just cleaning up some of the cases

Comment (wolfsage) [reply lib/Synergy/Reactor/HighFive.pm]: Ah no, it doesn't. Blarg.

**Comment by rjbs** [comment]
I think this LGTM, but!  Right now, we generate an SQLite db entry for each highfive.  I will give you the code.  This is used for the newsletter.  We need to keep that.  With that done, we'll cut over.

```perl
    my $dbh = DBI->connect(
      "dbi:SQLite:dbname=/home/rjbs/.synergy/synergy.sqlite",
      undef,
      undef,
    );

    $dbh->do(
      "INSERT INTO highfives
      (from_user, to_user, reason, highfived_at)
      VALUES
      (?, ?, ?, ?)",
      undef,
      $who,
      $target,
      $reason // "",
      time,
    );
```

**Pull request merged**